File: macos/SeedSeeker/Tests/SeedSeekerKitTests/ArtifactTests.swift
import Foundation
import XCTest
@testable import SeedSeekerKit

final class ArtifactTests: XCTestCase {
    func testScoutArtifactsShowRoundedGameLevels() {
        for item in ItemCatalog.artifacts {
            let expected = item.id == "sandals_of_nature" ? 7 :
                (["ethereal_chains", "timekeepers_hourglass"].contains(item.id) ? 6 : 5)
            let scout = ScoutItem(item: item, depth: 19, upgrade: 5, source: .impReward)
            XCTAssertEqual(scout.displayedUpgrade, expected)
            XCTAssertEqual(scout.upgrade, 5)
            XCTAssertEqual(ScoutItem(item: item, depth: 1, upgrade: 0, source: .heap).displayedUpgrade, 0)
        }
    }

    func testNamedArtifactsUseTheirOwnUpgradeCeilingAndCannotStack() throws {
        XCTAssertEqual(ItemKind.artifact.rawValue, 7)
        XCTAssertEqual(ItemCatalog.artifacts.count, 11)
        XCTAssertNil(ItemCatalog.findById("cloak_of_shadows"))
        XCTAssertNil(ItemCatalog.findById("holy_tome"))
        for item in ItemCatalog.artifacts {
            let requirement = try ItemRequirement(key: 1, item: item, upgrade: 5,
                kind: .artifact, upgradeMatch: .exactly, source: .impReward,
                maximumDepth: 19, requireUncursed: true)
            XCTAssertEqual(requirement.maximumUpgrade, 5)
            XCTAssertFalse([requirement].canStack([requirement].boardItems()[0]))
            XCTAssertThrowsError(try ItemRequirement(key: 2, item: item, upgrade: 6,
                kind: .artifact, upgradeMatch: .exactly))
            XCTAssertThrowsError(try ItemRequirement(key: 2, item: item, upgrade: 0,
                effect: .anyEnchantment, kind: .artifact, upgradeMatch: .any))
        }
        XCTAssertThrowsError(try ItemRequirement(key: 1, item: nil, upgrade: 0,
            kind: .artifact, upgradeMatch: .any))
    }

    func testRepeatedArtifactsRemainSeparateAndJoinWithoutUnnamedCopies() throws {
        let item = try XCTUnwrap(ItemCatalog.findById("ethereal_chains"))
        let first = try ItemRequirement(key: 1, item: item, upgrade: 0,
            kind: .artifact, upgradeMatch: .any)
        let second = try ItemRequirement(key: 2, item: item, upgrade: 0,
            kind: .artifact, upgradeMatch: .any, maximumDepth: 14)
        let requirements = [first, second]
        let board = requirements.boardItems()
        XCTAssertEqual(board.count, 2)
        XCTAssertTrue(board.allSatisfy { $0.extras.isEmpty && !requirements.canStack($0) })
        let joined = requirements.joinAlternatives(source: 0, target: 1)
        XCTAssertEqual(joined.count, 2)
        XCTAssertEqual(joined.boardCount, 1)
        XCTAssertEqual(joined.slotCount, 1)
        XCTAssertTrue(joined.allSatisfy { $0.item == item && $0.identityGroup == nil })
        XCTAssertNotNil(SavedQuery(requirements: joined).validated())
        XCTAssertNoThrow(try SearchRequest(requirements: joined))
    }

    func testArtifactAlternativesRoundTripThroughPersistenceDocumentsAndLinks() throws {
        let first = try ItemRequirement(key: 1,
            item: XCTUnwrap(ItemCatalog.findById("sandals_of_nature")), upgrade: 5,
            kind: .artifact, upgradeMatch: .exactly, source: .impReward,
            maximumDepth: 19, requireUncursed: true)
        let second = try ItemRequirement(key: 2,
            item: XCTUnwrap(ItemCatalog.findById("ethereal_chains")), upgrade: 0,
            kind: .artifact, upgradeMatch: .any, maximumDepth: 14)
        let requirements = [first, second].joinAlternatives(source: 0, target: 1)
        XCTAssertEqual(requirements.slotCount, 1)
        XCTAssertFalse(requirements.canStack(requirements.boardItems()[0]))
        let query = SavedQuery(requirements: requirements)
        let persisted = try JSONDecoder().decode(SavedQuery.self, from: JSONEncoder().encode(query))
        XCTAssertEqual(persisted, query)
        XCTAssertNotNil(persisted.validated())
        let restored = try ResultsExport.decodeQuery(ResultsExport.encodeQuery(query))
        let linked = try DeepLink.decode(DeepLink.encodeLink(for: query))
        for decoded in [restored, linked] {
            var expected = requirements
            var actual = decoded.requirements
            for index in expected.indices { expected[index].key = 0 }
            for index in actual.indices { actual[index].key = 0 }
            XCTAssertEqual(actual, expected)
            XCTAssertEqual(actual.slotCount, 1)
        }
    }

    func testNativeScoutVaultUpgradeAndMatchIndicesAgree() async throws {
        let world = try await ProductionSeedFinderEngine().scoutSeed("AAA-AAA-AAA", challenges: 0)
        let artifacts = world.items.filter { $0.item.kind == .artifact }
        XCTAssertEqual(artifacts.count, 4)
        let vault = try XCTUnwrap(artifacts.first { $0.source == .impReward })
        XCTAssertEqual(vault.item.id, "sandals_of_nature")
        XCTAssertEqual(vault.depth, 19)
        XCTAssertEqual(vault.upgrade, 5)
        XCTAssertFalse(vault.cursed)
        XCTAssertTrue(artifacts.filter { $0.source != .impReward }.allSatisfy { $0.upgrade == 0 })
        var requirement = try ItemRequirement(key: 1, item: vault.item, upgrade: 5,
            kind: .artifact, upgradeMatch: .exactly, source: .impReward,
            maximumDepth: 19, requireUncursed: true)
        let marks = try ScoutMatches.mark(seed: world.seed, challenges: 0,
            query: SearchRequest(requirements: [requirement]))
        XCTAssertEqual(marks.matchedRequirements, 1)
        let matched = world.items[try XCTUnwrap(marks.matched.first)]
        XCTAssertEqual(matched.item.id, vault.item.id)
        XCTAssertEqual(matched.upgrade, 5)
        requirement.maximumDepth = 18
        let tooEarly = try ScoutMatches.mark(seed: world.seed, challenges: 0,
            query: SearchRequest(requirements: [requirement]))
        XCTAssertEqual(tooEarly.matchedRequirements, 0)
    }

    @MainActor
    func testImportedArtifactProbabilityWaitsForSearchStatus() throws {
        let requirement = try ItemRequirement(key: 1,
            item: XCTUnwrap(ItemCatalog.findById("ethereal_chains")), upgrade: 0,
            kind: .artifact, upgradeMatch: .any)
        let controller = SearchController()
        controller.loadImported(seeds: ["AAA-AAA-AAA"], query: SavedQuery(requirements: [requirement]))
        XCTAssertEqual(controller.probabilityLabel, NumberFormat.probabilityPercent(nil))
        XCTAssertNil(controller.timeToSeed)
        XCTAssertFalse(controller.isImpossibleQuery)
    }

    func testArtifactProbabilityIsAvailableFromNativeStatus() async throws {
        let requirement = try ItemRequirement(key: 1,
            item: XCTUnwrap(ItemCatalog.findById("ethereal_chains")), upgrade: 0,
            kind: .artifact, upgradeMatch: .any)
        let session = try await ProductionSeedFinderEngine().startSearch(
            SearchRequest(requirements: [requirement]), workers: 1)
        do {
            let status = try await session.status()
            XCTAssertTrue(status.matchProbability.isFinite)
            XCTAssertGreaterThan(status.matchProbability, 0)
            XCTAssertLessThan(status.matchProbability, 1)
            XCTAssertEqual(status.state, .running)
        } catch {
            await session.cancel()
            await session.close()
            throw error
        }
        await session.cancel()
        await session.close()
    }
}

File: windows/SeedSeeker.Tests/ArtifactTests.cs
using System.Text.Json;
using Xunit;

namespace SeedSeeker.Tests;

public sealed class ArtifactTests
{
    [Fact]
    public void ScoutArtifactsShowRoundedGameLevels()
    {
        foreach (var item in ItemCatalog.For(ItemKind.Artifact))
        {
            var expected = item.Id switch { "sandals_of_nature" => 7,
                "ethereal_chains" or "timekeepers_hourglass" => 6, _ => 5 };
            var scout = new ScoutItem(item, 19, 5, null, false, ScoutItemSource.ImpReward, 0, 0, 0);
            Assert.Equal(expected, scout.DisplayedUpgrade);
            Assert.Equal(5, scout.Upgrade);
            Assert.Equal(0, (scout with { Upgrade = 0 }).DisplayedUpgrade);
        }
    }

    private static ItemRequirement Sandals() => new()
    {
        Kind = ItemKind.Artifact, Item = ItemCatalog.Find("sandals_of_nature"),
        UpgradeMatch = UpgradeMatch.Exactly, Upgrade = 5,
        Source = ScoutItemSource.ImpReward, MaximumDepth = 19, RequireUncursed = true,
    };

    [Fact]
    public void ArtifactCatalogAndEditorBoundsMatchNamedArtifacts()
    {
        var items = ItemCatalog.For(ItemKind.Artifact).ToList();
        Assert.Equal(11, items.Count);
        Assert.Equal(11, items.Select(item => item.Id).Distinct().Count());
        Assert.All(items, item => Assert.Null(item.Tier));
        Assert.True(ItemKind.Artifact.RequiresNamedItem());
        Assert.Equal(5, Sandals().UpgradeCeiling);
        Assert.Empty(ItemCatalog.Modifiers(ItemKind.Artifact));
        Assert.Contains("+5", Sandals().Description);
        Assert.Contains("by floor 19", Sandals().Description);
    }

    [Fact]
    public void ArtifactsCannotBeWildcardsOrStacks()
    {
        var unnamed = new QuerySettings { Requirements = [new() { Kind = ItemKind.Artifact }] };
        Assert.Equal("Choose a named artifact.", QueryRelationships.Validate(unnamed));
        Assert.Null(NativeEngine.TryEncodeShareLink(ResultsExport.EncodeQueryDocument(unnamed)));
        var query = new QuerySettings { Requirements = [Sandals()] };
        var item = Assert.Single(QueryRelationships.BoardItems(query.Requirements));
        Assert.False(QueryRelationships.CanStack(query.Requirements, item));
        Assert.Single(QueryRelationships.SetStackCount(query.Requirements, item, 3));
        var plain = new ItemRequirement { Kind = ItemKind.Artifact, Item = ItemCatalog.Find("dried_rose") };
        var repeated = new[] { plain, plain.Clone(), Sandals() };
        Assert.Equal(3, QueryRelationships.BoardItems(repeated).Count);
        var joined = QueryRelationships.JoinAlternatives(repeated, 0, 2);
        Assert.All(joined, requirement => Assert.Null(requirement.IdentityGroup));
    }

    [Fact]
    public void ArtifactConstraintsAndAlternativesSurviveDocumentsLinksAndSettings()
    {
        var alternative = new ItemRequirement { Kind = ItemKind.Artifact, Item = ItemCatalog.Find("dried_rose"), MaximumDepth = 9 };
        var query = new QuerySettings { Requirements = new(QueryRelationships.JoinAlternatives([Sandals(), alternative], 1, 0)) };
        var document = ResultsExport.EncodeQueryDocument(query);
        Assert.Contains("\"kind\":\"artifact\"", document);
        var link = NativeEngine.TryEncodeShareLink(document);
        Assert.NotNull(link);
        var linkDocument = NativeEngine.TryDecodeShareText(link!);
        Assert.NotNull(linkDocument);
        var restored = ResultsExport.DecodeQueryDocument(linkDocument!);
        Assert.Equal(1, QueryRelationships.SlotCount(restored.Requirements));
        Assert.All(restored.Requirements, requirement => Assert.Equal(ItemKind.Artifact, requirement.Kind));
        var sandals = restored.Requirements.Single(requirement => requirement.Item?.Id == "sandals_of_nature");
        Assert.Equal(5, sandals.Upgrade);
        Assert.Equal(19, sandals.MaximumDepth);
        Assert.True(sandals.RequireUncursed);
        Assert.Equal(ScoutItemSource.ImpReward, sandals.Source);
        var file = ResultsExport.Encode(query, ["AAA-AAA-AAA"], "test");
        Assert.Equal(document, ResultsExport.EncodeQueryDocument(ResultsExport.Decode(file).Query));
        var settings = JsonSerializer.Deserialize<QuerySettings>(JsonSerializer.Serialize(query))!;
        Assert.Equal(document, ResultsExport.EncodeQueryDocument(settings));
    }

    [Fact]
    public void NativeScoutKeepsArtifactUpgradesAndMatchIndices()
    {
        var world = new NativeEngine().Scout("AAA-AAA-AAA", 0);
        Assert.Equal(4, world.Items.Count(item => item.Item.Kind == ItemKind.Artifact));
        var sandals = Assert.Single(world.Items, item => item.Item.Id == "sandals_of_nature");
        Assert.Equal(5, sandals.Upgrade);
        Assert.Equal(19, sandals.Depth);
        Assert.Equal(ScoutItemSource.ImpReward, sandals.Source);
        var query = new QuerySettings { Requirements = [Sandals()] };
        var matches = NativeEngine.ScoutMatches(world.Seed, 0, query);
        Assert.Equal(1, matches.MatchedRequirements);
        Assert.Equal(sandals, world.Items[Assert.Single(matches.Matched)]);
        query.Requirements[0].MaximumDepth = 18;
        Assert.Empty(NativeEngine.ScoutMatches(world.Seed, 0, query).Matched);
    }

    [Theory]
    [InlineData(double.NaN)]
    [InlineData(double.PositiveInfinity)]
    public void UnavailableProbabilityHasAStableDisplay(double probability)
    {
        var status = new SearchStatus(SearchState.Running, 1, 10, 0, probability);
        Assert.True(status.ProbabilityUnavailable);
        Assert.Equal("unavailable", status.ProbabilityDescription);
    }
}
